app/libpriv: add editor option for karg command

peterbaouoft opened the pull request:
Add editor option for kernel arguments. The logic
was taken from the ostree-commit editor. The editor code is also
taken from ot-editor.c.

The user now can edit the kernel arguments from an editor.

The synchronization related issue is resolved by reloading the
sysroot and see if the sysroot has been changed since the call
of the editor prompt.

cgwalters commented on src/app/rpmostree-builtin-kargs.c:
```diff
@@ -62,6 +59,104 @@ get_kargs_option_variant (void)
+      "# should be in the form of key=value.\n"
+      "# Lines starting with '#' will be ignored. Each key=value pair should be \n"
+      "# separated by spaces, and multiple value associated with one key is allowed. \n"
+      "# Also, please note that any changes to existing ostree argument will not be \n"
```
Let's ensure we omit the `ostree=` argument in this, and error out if it's provided?

peterbaouoft replied:
sure, added in the fixup below. Any appearance of `ostree` from the user input will error out now.

cgwalters commented on src/app/rpmostree-builtin-kargs.c:
```diff
@@ -62,6 +59,104 @@ get_kargs_option_variant (void)
+  /* We now process the editor output, we ignore lines starting with #
+   * and we assume the kernel arguments are usually in one line.
+   */
+  g_auto(GStrv) lines = g_strsplit(out_editor_string, "\n", -1);
```
style nit, space between identifier and paren

cgwalters commented on src/app/rpmostree-builtin-kargs.c:
```diff
@@ -62,6 +59,104 @@ get_kargs_option_variant (void)
+      g_strchomp (lines[i]);
+      g_strchug (lines[i]);
+      /* comments and blank lines are skipped */
+      if (lines[i][0] == '#')
```
This would read a bit more nicely I think if we did `char *line = lines[i];` at the top, then `*line == '#'` or so.

cgwalters commented on src/app/rpmostree-builtin-kargs.c:
```diff
@@ -62,6 +59,104 @@ get_kargs_option_variant (void)
+       */
+      g_string_append (kernel_arg_buf, " ");
+    }
+  g_strchomp (kernel_arg_buf->str);
```
Rather than doing this we could do:

`if (i < num_lines-1) g_string_append (kernel_arg_buf, " ");`

peterbaouoft replied:
hmm, I tried that earlier, but this will result into some extra spaces in the case of 
```
argument
\blank line
\blank line 
```
then, the space will always be added even though other lines are skipped (I think). Then due to the added space, the comparison check later for string equal will fail.. 

It also seems like the editor tends to automatically add a new line to the file at the end, whenever I use `wq` to exit the editor. I tested it a bit and found some documents online about this behaviour. 
https://stackoverflow.com/questions/14171254/why-would-vim-add-a-new-line-at-the-end-of-a-file 
https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline.


So I think we may need to consider the "added new line (by editor)" case, so I think we could either do 
`if ( i < num_lines - 2) g_string_append (kernel_arg_buf, " ");` Or we chomp the output directly?

cgwalters replied:
Don't worry about it then, let's keep it as is!

peterbaouoft replied:
ouch, just noticed this comment :-(, lemme try again then. :p

cgwalters commented on src/app/rpmostree-builtin-kargs.c:
```diff
@@ -62,6 +59,104 @@ get_kargs_option_variant (void)
+    }
+  g_strchomp (kernel_arg_buf->str);
+  /* We do not allow empty kernel arg string */
+  if (!strcmp (kernel_arg_buf->str, ""))
```
Let's use `if (g_str_equal (kernel_arg_buf->str, ""))` ?

cgwalters commented on src/app/rpmostree-builtin-kargs.c:
```diff
@@ -160,42 +253,90 @@ rpmostree_ex_builtin_kargs (int            argc,
+      if (out_changed)
+        {
+          g_set_error (error, G_IO_ERROR, G_IO_ERROR_INVALID_ARGUMENT,
+                       "Bootconfig has already been changed since then, please retry the editor");
```
How about "Conflict: bootloader configuration changed"?

cgwalters replied:
Also IMO we should avoid deleting the temp file in this case.  See for example what I [do in git-evtag](https://github.com/cgwalters/git-evtag/blob/d32fdef6694d1a47965c6d5c1c50b8d2bf87f95e/src/git-evtag.c#L725).  (I *had* to add this as sometimes my Yubikey Neo isn't initialized, and I'd write a long git tag message only to have it be discarded when `gpg2` failed to sign it)

peterbaouoft replied:
sure, now in the fixup, we tend to rewrite user's input into a new temp file created. The reason for doing that is so we do not need to change the original code copied from `ot-editor.c`. 

A cleaner way(I think) would be to directly modify the editor function copied from `ot-editor.c` to also return the file (and not deleting it).  but yea... change shared code. seems not that neat (I feel). 

So, If there is other easier implementations, please let me know =)

peterbaouoft commented:
Many thanks for the review =) and added a fixup ⬆️ and addressed comments. 

There is also one thing I noticed: we tend to not care about `ostree` argument in the editor code. But I think we have not done that on the server side (delete, append,replace). Should we also take care of that (ignore ostree and error out)? 

If we also want to do that, I think of two options for now: (but both seems .... not nice to implement)
1:  we check each of the input  (delete, append ,and replace). If any of the input contain a string with ostree as the prefix, we error out. (that does require.. many looping I think)
2: Or we omit the ostree argument before passing into dbus. then on the dbus side, before we write the arguments to the sysroot, we can perform a check to see if ostree argument exist. If it does exist, we can error out on the dbus side?

Any suggestions would be great to have, and thanks for your time =).

jlebon commented on src/app/rpmostree-builtin-kargs.c:
```diff
@@ -62,6 +60,110 @@ get_kargs_option_variant (void)
+      g_string_append (kernel_arg_buf,  " ");
+    }
+
+  g_strchomp (kernel_arg_buf->str);
```
Hmm, this is modifying the `GString`-managed buffer directly though. E.g. the `->len` argument will now be inconsistent, right? Maybe let's just `g_string_free` from here to take ownership and then we can `g_strchomp`?

peterbaouoft replied:
sure =)

jlebon commented on src/app/rpmostree-builtin-kargs.c:
```diff
@@ -62,6 +60,110 @@ get_kargs_option_variant (void)
+      char *line = lines[i];
+      /* Remove leading and trailing spaces */
+      g_strchomp (line);
+      g_strchug (line);
```
`g_strstrip` ? :)

jlebon commented on src/app/rpmostree-builtin-kargs.c:
```diff
@@ -62,6 +60,110 @@ get_kargs_option_variant (void)
+
+  g_strchomp (kernel_arg_buf->str);
+
+  if (g_strrstr (kernel_arg_buf->str, "ostree"))
```
Doesn't this mean that e.g. `my_random_key="string with ostree word in it"` will fail? Maybe let's just reparse the kernel arguments into key val pairs and explicitly look for the `ostree` key?

peterbaouoft replied:
yup, sounds good, I did not bother for the first time, but I think it is worth to do it :p.

jlebon commented on src/app/rpmostree-builtin-kargs.c:
```diff
@@ -84,18 +186,16 @@ rpmostree_ex_builtin_kargs (int            argc,
+       * Thus erroring out ahead of time when these strings exist
+       */
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_INVALID_ARGUMENT,
+                   "Editor command is already enough to modify the kernel arguments");
```
Let's be more explicit here, e.g. `"Cannot specify --editor with --replace, --delete, or --append"` ?

peterbaouoft replied:
Sure =)

jlebon commented on src/app/rpmostree-builtin-kargs.c:
```diff
@@ -160,42 +260,108 @@ rpmostree_ex_builtin_kargs (int            argc,
 
   if (display_kernel_args)
     {
       g_print("The kernel arguments are:\n%s\n",
```
Minor: missing space before opening parenthesis. :)

peterbaouoft replied:
yup, nice catch!

jlebon commented on src/app/rpmostree-builtin-kargs.c:
```diff
@@ -160,42 +260,108 @@ rpmostree_ex_builtin_kargs (int            argc,
+        return EXIT_FAILURE;
+      if (out_changed)
+        {
+          /* We store user's input into a file, and notify the user that bootconfig has already been changed
```
Since it shouldn't be that long, maybe let's just skip writing to a file and just dump it on the terminal? E.g. make the error we throw `Conflict: bootloader configuration changed`, but before returning print something like

```
Saved kernel arguments:\n%s\n
```

?

peterbaouoft replied:
yup, that sounds a lot nicer  :D, thanks for the suggestion. Tho the last "\n" seems like needed to be removed? As we print an extra line for error in main.c? https://github.com/projectatomic/rpm-ostree/blob/master/src/app/main.c#L432

jlebon replied:
Sure, that's fine!

peterbaouoft commented on src/app/rpmostree-builtin-kargs.c:
```diff
@@ -45,9 +45,7 @@ static GOptionEntry option_entries[] = {
-  /* TODO: Add back the editor option once we decided how to handle the similar API from ostree
-   * { "editor", 0, 0, G_OPTION_ARG_NONE, &opt_editor, "Pops up an editor for users to change arguments", NULL },
-   */
+  { "editor", 0, 0, G_OPTION_ARG_NONE, &opt_editor, "Pops up an editor for users to change kernel arguments", NULL },
```
sure =)

peterbaouoft commented:
Thanks for the review. Pushed another fixup ⬆️based on the comments =).

jlebon commented on src/app/rpmostree-builtin-kargs.c:
```diff
@@ -84,18 +194,16 @@ rpmostree_ex_builtin_kargs (int            argc,
+       * Thus erroring out ahead of time when these strings exist
+       */
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_INVALID_ARGUMENT,
+                   "Cannot specify --editor with --replace --delete, or --append");
```
Minor: missing comma after `--replace`.

peterbaouoft replied:
yup!

jlebon commented on src/app/rpmostree-builtin-kargs.c:
```diff
@@ -62,6 +60,118 @@ get_kargs_option_variant (void)
+  g_autofree char *kernel_args_str = g_string_free (g_steal_pointer(&kernel_arg_buf), FALSE);
+  g_strchomp (kernel_args_str);
+
+  g_auto(GStrv) kernel_arg_list = g_strsplit (kernel_args_str, " ", -1);
```
What I meant here was to just use `_ostree_kernel_args_from_string` again and then just query using `_ostree_kernel_args_get_last_value` like we do above?

peterbaouoft replied:
ah I see =)

jlebon commented on src/app/rpmostree-builtin-kargs.c:
```diff
@@ -160,42 +268,91 @@ rpmostree_ex_builtin_kargs (int            argc,
+      if (out_changed)
+        {
+          g_set_error (error, G_IO_ERROR, G_IO_ERROR_INVALID_ARGUMENT,
+                       "Conflict: bootloader configuration changed, Saved kernel arguments: \n%s", current_kernel_arg_string);
```
This is OK, though it feels cleaner to just `g_print` the saved arguments instead of making it part of the error message? Hmm not sure. At least `s/,/./` though :)

peterbaouoft replied:
sure =).

jlebon commented:
Nice, that looks good to me!
I'll let @cgwalters have a final look as well.

cgwalters commented:
Looks pretty good to me, thanks!

@rh-atomic-bot r=jlebon 549e8f5

peterbaouoft commented:
Thanks for the review =)
